src/components/ble/AppleNotificationCenterClient.cpp

```diff
@@ -0,0 +1,514 @@
+    }
+
+    if (notifications.contains(notificationUuid)) {
+      notifications[notificationUuid] = ancsNotif;
+    } else if (!silent) {
+      notifications.insert({notificationUuid, ancsNotif});
+    }
+
+    if (preExisting || eventId != static_cast<uint8_t>(EventIds::Added) || silent) {
+      return;
+    }
+
+    // Request ANCS more info
+    uint8_t titleSize = maxTitleSize + 4;
+    uint8_t subTitleSize = maxSubtitleSize + 4;
+    uint8_t messageSize = maxMessageSize + 4;
+    BYTE request[14];
+    request[0] = 0x00; // Command ID: Get Notification Attributes
+    request[1] = (uint8_t) (notificationUuid & 0xFF);
+    request[2] = (uint8_t) ((notificationUuid >> 8) & 0xFF);
+    request[3] = (uint8_t) ((notificationUuid >> 16) & 0xFF);
+    request[4] = (uint8_t) ((notificationUuid >> 24) & 0xFF);
+    request[5] = 0x01; // Attribute ID: Title
+    // request[6] = 0x00;
+    request[6] = (titleSize & 0xFF);
+    request[7] = ((titleSize >> 8) & 0xFF);
+    request[8] = 0x02; // Attribute ID: Subtitle
+    request[9] = (subTitleSize & 0xFF);
+    request[10] = ((subTitleSize >> 8) & 0xFF);
+    request[11] = 0x03; // Attribute ID: Message
+    request[12] = (messageSize & 0xFF);
+    request[13] = ((messageSize >> 8) & 0xFF);
+
+    ble_gattc_write_flat(event->notify_rx.conn_handle, controlPointHandle, request, sizeof(request), OnControlPointWriteCallback, this);
+  } else if (event->notify_rx.attr_handle == dataSourceHandle || event->notify_rx.attr_handle == dataSourceDescriptorHandle) {
+    uint16_t titleSize;
+    uint16_t subTitleSize;
+    uint16_t messageSize;
+    uint32_t notificationUid;
+
+    os_mbuf_copydata(event->notify_rx.om, 1, 4, &notificationUid);
+    os_mbuf_copydata(event->notify_rx.om, 6, 2, &titleSize);
+    os_mbuf_copydata(event->notify_rx.om, 8 + titleSize + 1, 2, &subTitleSize);
+    os_mbuf_copydata(event->notify_rx.om, 8 + titleSize + 1 + 2 + subTitleSize + 1, 2, &messageSize);
+
+    AncsNotification ancsNotif;
+    ancsNotif.uuid = 0;
+
+    if (notifications.contains(notificationUid)) {
+      ancsNotif = notifications[notificationUid];
+    }
+
+    std::string decodedTitle = DecodeUtf8String(event->notify_rx.om, titleSize, 8);
+
+    std::string decodedSubTitle = DecodeUtf8String(event->notify_rx.om, subTitleSize, 8 + titleSize + 1 + 2);
+
+    std::string decodedMessage = DecodeUtf8String(event->notify_rx.om, messageSize, 8 + titleSize + 1 + 2 + subTitleSize + 1 + 2);
+
+    NRF_LOG_INFO("Decoded Title: %s", decodedTitle.c_str());
+    NRF_LOG_INFO("Decoded SubTitle: %s", decodedSubTitle.c_str());
+
+    bool incomingCall = ancsNotif.uuid != 0 && ancsNotif.category == static_cast<uint8_t>(Categories::IncomingCall);
+
+    if (!incomingCall) {
+      if (titleSize >= maxTitleSize) {
+        decodedTitle.resize(maxTitleSize - 3);
+        decodedTitle += "...";
+        if (!decodedSubTitle.empty()) {
+          decodedTitle += " - ";
+        } else {
+          decodedTitle += "-";
+        }
```

Comment:
if `decodedTitle` is longer than `maxTitleSize`, then you resize it to 3 chars less than max. Then you add `"..."`, which makes it full again, and then you add `" - "` or even `-`. To me, it seems like `decodedTitle` is again too long after that.